pkg/storage/storage.go

```diff
@@ -320,6 +320,58 @@ func (s *Storage) Get(gi *GetInput) (*GetOutput, error) {
 	}, nil
 }
 
+type DeleteInput struct {
+	StartTime time.Time
+	EndTime   time.Time
+	Key       *Key
+}
+
+func (s *Storage) Delete(di *DeleteInput) error {
+	s.closingMutex.Lock()
+	defer s.closingMutex.Unlock()
+
+	if s.closing {
```

Comment:
It is not related to the change itself but I'm wondering if we really need to keep `closingMutex` acquired during the whole call. Do other functions actually compete? If so, should we rename it accordingly (just `m`/`mu`)?

Reply:
Yeah, this is a very good point. The idea was to make sure we don't try to read / write after `Close` is called.

But the way it is currently implemented, we basically have a lock on the whole db... I changed it to a `RWMutex`. 

Good news is that this means we'll get better performance. Bad news is that this might now expose some bugs that were hidden before because we had a lock on the whole db. Hopefully unit tests + running pyroscope server locally will expose those.

Reply:
Oh, that's my biggest concern. We can mitigate the risk by using `RLock` in `Get` only.

Also, I think `Close` should acquire the mutex for the whole call.

Reply:
If Close acquires mutex for the whole call, then Get and Put will just block instead of returning errors... I think I'd rather them returning errors..

> We can mitigate the risk by using RLock in Get only.

I'll do some tests today. One thing I can tell already is that if I [benchmark it](https://github.com/pyroscope-io/pyroscope/tree/main/benchmark) nothing breaks. That's good, but is not a 100% guarantee.

Reply:
I was wondering also about a situation when we remove from cache, but not yet from disk. If in the middle someone would call a read, cache would read from disk, reloading an item that was previously deleted. And at the end, we will in fact remove from disk, but we end up with not removed item in cache.